amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageReader.java

```diff
@@ -0,0 +1,127 @@
+     * @param payload                  The message payload or {@code null} if the message does not contain any payload.
+     * 
+     * @throws NullPointerException if the applicationPropereties is {@code null}.
+     */
+    protected AmqpBinaryMessageReader(final SpecVersion version, final ApplicationProperties applicationPropereties, 
+            final String contentType, final byte[] payload) {
+        super(version, payload != null && payload.length > 0 ? new BytesCloudEventData(payload) : null);
+        this.contentType = contentType;
+        this.applicationPropereties = Objects.requireNonNull(applicationPropereties);
+    }
+
+    @Override
+    protected boolean isContentTypeHeader(final String key) {
+        return key.equals(AmqpConstants.PROPERTY_CONTENT_TYPE);
+    }
+
+    /**
+     * Tests whether the given attribute key is prefixed with <em>cloudEvents:</em>
+     * 
+     * @param key   The key to test for the presence of the prefix.
+     * @return      True if the specified key starts with the prefix or
+     *              false otherwise.
+     */
+    @Override
+    protected boolean isCloudEventsHeader(final String key) {
+        final int prefixLength = AmqpConstants.CE_PREFIX.length();
+        return key.length() > prefixLength && key.startsWith(AmqpConstants.CE_PREFIX);
+    }
+
+    /**
+     * Gets the cloud event attribute key without the preceding prefix.
+     * 
+     * @param key  The key containing the AMQP specific prefix.
+     * 
+     * @return     The key without the prefix.
+     */
+    @Override
+    protected String toCloudEventsKey(final String key) {
+        return key.substring(AmqpConstants.CE_PREFIX.length());
+    }
+
+    /**
+     * Visits the <em>content-type</em> message property and all <em>application-properties</em> 
+     * of this message reader.
+     * 
+     * @param  fn A callback to consume this reader's application-properties
+     *            and content-type property.
+     */
+    @Override
+    protected void forEachHeader(final BiConsumer<String, Object> fn) {
+        if (contentType != null) {
+            // visit the content-type message property
+            fn.accept(AmqpConstants.PROPERTY_CONTENT_TYPE, contentType);
+        }
+        // visit application-properties
+        applicationPropereties.getValue().forEach((k, v) -> fn.accept(k, v));
+    }
+
+    /**
+     * Gets the cloud event representation of the value.
+     * <p>
+     * This method simply returns the string representation of the type of value passed as argument.
+     * 
+     * @param value The value of a CloudEvent attribute or extension.
+     * @return The string representation of the specified value.
+     * 
+     * @throws NullpointerException if the value is {@code null}.
+     */
+    @Override
+    protected String toCloudEventsValue(final Object value) {
+        return Objects.requireNonNull(value).toString();
```

Comment:
Can the attribute be there in the applicationProperties map but with a null value?

Reply:
i think this depends on how strict the map implementation used for the applicationProperties is. This reader uses the available HashMap implementation which is less strict and thus allows `null` values (and even null keys). since the `ApplicationProperties` is defined to take any Map type, i added this just to be on the safe side

but i can check if there is already a stricter map implementation for app properties present within proton-j

Reply:
I think it's fine, my only concern is that, if the value is null, then this method shouldn't be called at all. Maybe you need the null check in `forEachHeader`?